### PR TITLE
Create new objects with id set when saving.

### PR DESCRIPTION
**Pull request by meistermeier**

### CODE REVIEW

File: src/main/java/org/springframework/data/neo4j/core/ReactiveNeo4jTemplate.java [hunk 1]
```diff
@@ -562,140 +566,224 @@ private Object convertIdValues(@Nullable Neo4jPersistentProperty idProperty, Obj
...
+			Flux<RelationshipPropertyValues> relationshipCreation = Flux.fromIterable(relatedValuesToStore).flatMap(relatedValueToStore -> {
+
+				Object relatedNodePreEvt = relationshipContext.identifyAndExtractRelationshipTargetNode(relatedValueToStore);
+				return Mono.deferContextual(ctx -> eventSupport.maybeCallBeforeBind(relatedNodePreEvt)
+						.flatMap(relatedNode -> Mono.just(neo4jMappingContext.getPersistentEntity(relatedNodePreEvt.getClass()))
+								.flatMap(targetEntity -> Mono.just(targetEntity.isNew(relatedNode))
+										.flatMap(isNew -> saveRelatedNode(relatedNode, relationshipContext.getAssociationTargetType(),
+												targetEntity)
+												.flatMap(relatedInternalId -> {
+													// if an internal id is used this must get set to link this entity in the next iteration
```
Comment (michael-simons) [inline src/main/java/org/springframework/data/neo4j/core/ReactiveNeo4jTemplate.java:657]: > get set

?

File: src/main/java/org/springframework/data/neo4j/core/ReactiveNeo4jTemplate.java [hunk 2]
```diff
@@ -562,140 +566,224 @@ private Object convertIdValues(@Nullable Neo4jPersistentProperty idProperty, Obj
...
+																if (processState != ProcessState.PROCESSED_ALL_VALUES) {
+																	something = processNestedRelations(targetEntity, targetPropertyAccessor,
+																			isNew, stateMachine);
+																} else {
+																	something = Mono.just((T) targetPropertyAccessor.getBean());
+																}
+
+																return something.flatMap(currentValue -> {
+																	Object newRelationshipObject = targetPropertyAccessor.getBean();
+																	if (relationshipDescription.hasRelationshipProperties()) {
```
Comment (michael-simons) [inline src/main/java/org/springframework/data/neo4j/core/ReactiveNeo4jTemplate.java:694]: Interesting choice for a variable name.

File: src/main/java/org/springframework/data/neo4j/core/ReactiveNeo4jTemplate.java [hunk 3]
```diff
@@ -562,140 +566,224 @@ private Object convertIdValues(@Nullable Neo4jPersistentProperty idProperty, Obj
...
+																		Object lalelu = relationshipProperty.isDynamicAssociation()
```
Comment (michael-simons) [inline src/main/java/org/springframework/data/neo4j/core/ReactiveNeo4jTemplate.java:695]: I would rather prefer to bite it and deduplicate this conditions, similar on what we did together with the checks for cyclic dependencies, regardless of the fact that one lives in the imperative world and the other in the reactive world.

File: src/test/java/org/springframework/data/neo4j/integration/imperative/RepositoryIT.java [hunk 4]
```diff
@@ -1802,20 +1801,6 @@ void createWithCustomQueryShouldWorkWithPlainObjects(@Autowired PersonRepository
 			assertThat(newPerson.getBornOn()).isEqualTo(p.getBornOn());
 		}
 
-		@Test // GH-2141
-		void saveWithGeneratedIdsReturnsObjectWithIdSet(
```
Comment (michael-simons) [inline src/test/java/org/springframework/data/neo4j/integration/imperative/RepositoryIT.java:1806]: I appreciate that removal a lot.

**Review by michael-simons — CHANGES_REQUESTED** [review]
In general I am wondering if there's a way to check upfront wether the recreation of the collections is actually necessary (at least if I don't understand the changes wrong, the templates seem to create at least the new approximate collection in all cases, regardless of whether working with immutable or not.).

**Comment by meistermeier** [comment]
Merged with f9267d35a

**Pull request closed without merging**